lib/vrl/stdlib/src/unnest.rs

```diff
@@ -0,0 +1,160 @@
+                    r#"[{"hostname": "localhost", "events": {"message": "hello"}}, {"hostname": "localhost", "events": {"message": "world"}}]"#,
+                ),
+            },
+            Example {
+                title: "variable target",
+                source: indoc! {r#"
+                    foo = {"hostname": "localhost", "events": [{"message": "hello"}, {"message": "world"}]}
+                    foo = unnest!(foo.events)
+                "#},
+                result: Ok(
+                    r#"[{"hostname": "localhost", "events": {"message": "hello"}}, {"hostname": "localhost", "events": {"message": "world"}}]"#,
+                ),
+            },
+        ]
+    }
+
+    fn compile(&self, mut arguments: ArgumentList) -> Compiled {
+        let path = arguments.required_query("path")?;
+
+        Ok(Box::new(UnnestFn { path }))
+    }
+}
+
+#[derive(Debug, Clone)]
+struct UnnestFn {
+    path: expression::Query,
+}
+
+impl UnnestFn {
+    #[cfg(test)]
+    fn new(path: &str) -> Self {
+        use std::str::FromStr;
+
+        Self {
+            path: expression::Query::new(
+                expression::Target::External,
+                FromStr::from_str(path).unwrap(),
+            ),
+        }
+    }
+}
+
+impl Expression for UnnestFn {
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        let path = self.path.path();
+
+        let value: Value;
+        let target: Box<&dyn Target> = match self.path.target() {
+            expression::Target::External => Box::new(ctx.target()) as Box<_>,
+            expression::Target::Internal(v) => {
+                let v = ctx.state().variable(v.ident()).unwrap_or(&Value::Null);
+                Box::new(v as &dyn Target) as Box<_>
+            }
+            expression::Target::Container(expr) => {
+                value = expr.resolve(ctx)?;
+                Box::new(&value as &dyn Target) as Box<&dyn Target>
+            }
+            expression::Target::FunctionCall(expr) => {
+                value = expr.resolve(ctx)?;
+                Box::new(&value as &dyn Target) as Box<&dyn Target>
+            }
+        };
+
+        let root = target.get(&LookupBuf::root())?.unwrap_or(Value::Null);
+
+        let values = root
+            .get_by_path(path)
+            .cloned()
+            .ok_or(value::Error::Expected {
+                got: Kind::Null,
+                expected: Kind::Array,
+            })?
+            .try_array()?;
+
+        let events = values
+            .into_iter()
+            .map(|value| {
+                let mut event = root.clone();
+                event.insert_by_path(path, value);
+                event
+            })
+            .collect::<Vec<_>>();
+
+        Ok(Value::Array(events))
+    }
+
+    fn type_def(&self, state: &state::Compiler) -> TypeDef {
+        self.path
+            .type_def(state)
+            .fallible_unless(Kind::Object)
+            .restrict_array()
+            .add_null()
+    }
```

Comment:
Given that we set the fallibility and type here, we don't need to base this off the `path` type def.

I'm also a bit confused what the `fallible_unless` is for? It doesn't appear to match the expectation that the path always points to an array?

I _believe_ the following should suffice, but correct me if I'm wrong!

```suggestion
        TypeDef::new()
            .infallible()
            .restrict_array()
            .add_null()
    }
```

Reply:
Aha, yes, you are right, I meant to do `Kind::Array` there. This function is always fallible, given we can't be sure that the provided path actually points to an array (or can we? It didn't seem like we could) but I agree it doesn't need to be based on the path. Will update.